### PR TITLE
Extend InstrumentationScopeInfo with scope attributes

### PR DESCRIPTION
**Pull request by jack-berg**
Related to #4695.

### CODE REVIEW

File: sdk/common/src/main/java/io/opentelemetry/sdk/internal/ComponentRegistry.java
Comment (breedx-splk) [inline sdk/common/src/main/java/io/opentelemetry/sdk/internal/ComponentRegistry.java:45]: The addition of the attributes param makes it a new method and invalidates the `@since` right?
```suggestion
   * @since 1.18.0
```
Internal, so probably not terribly important.

File: sdk/common/src/test/java/io/opentelemetry/sdk/internal/ComponentRegistryTest.java
Comment (breedx-splk) [inline sdk/common/src/test/java/io/opentelemetry/sdk/internal/ComponentRegistryTest.java:25]:
```suggestion
  void get_NameAndAttributesMustNotBeNull() {
```

Comment (breedx-splk) [reply sdk/common/src/test/java/io/opentelemetry/sdk/internal/ComponentRegistryTest.java:25]: Why have these tests that do double duty? I'd say let's split these up into separate small tests.

File: sdk/common/src/test/java/io/opentelemetry/sdk/internal/ComponentRegistryTest.java
Comment (breedx-splk) [inline sdk/common/src/test/java/io/opentelemetry/sdk/internal/ComponentRegistryTest.java:36]: Split to separate tests?

**Review by breedx-splk — COMMENTED** [review]
A few small suggestions but looks good to me. 👍🏻

File: sdk/common/src/main/java/io/opentelemetry/sdk/common/InstrumentationScopeInfo.java
Comment (jkwatson) [inline sdk/common/src/main/java/io/opentelemetry/sdk/common/InstrumentationScopeInfo.java:45]: we might be at the point where we introduce a builder for this fella

**Review by kubawach — APPROVED** [review]

File: sdk/common/src/test/java/io/opentelemetry/sdk/internal/ComponentRegistryTest.java
Comment (jack-berg) [reply sdk/common/src/test/java/io/opentelemetry/sdk/internal/ComponentRegistryTest.java:25]: The double duty is really just the set of invalid argument test cases. There's a fair amount of precedence for grouping invalid args tests together in this repo (for [example](https://github.com/open-telemetry/opentelemetry-java/blob/main/exporters/otlp/testing-internal/src/main/java/io/opentelemetry/exporter/otlp/testing/internal/AbstractGrpcTelemetryExporterTest.java#L601-L632)).

I think I'll keep the tests together but rename the test to be `get_InvalidArgs()`.

File: sdk/common/src/main/java/io/opentelemetry/sdk/internal/ComponentRegistry.java
Comment (jack-berg) [reply sdk/common/src/main/java/io/opentelemetry/sdk/internal/ComponentRegistry.java:45]: Yeah the `@since` is only required for methods part of the public api. I'll just remove the annotation altogether..

File: sdk/common/src/main/java/io/opentelemetry/sdk/common/InstrumentationScopeInfo.java
Comment (jack-berg) [reply sdk/common/src/main/java/io/opentelemetry/sdk/common/InstrumentationScopeInfo.java:45]: For me it was a question of likelihood that additional fields are added to instrumentation scope in the future. I think its unlikely, but it could still happen so probably best to introduce a builder. 

How about I add a builder and deprecate the the methods that accept version and schema?

File: sdk/common/src/test/java/io/opentelemetry/sdk/internal/ComponentRegistryTest.java
Comment (breedx-splk) [reply sdk/common/src/test/java/io/opentelemetry/sdk/internal/ComponentRegistryTest.java:25]: Yeah ok, that helps, and I understand the precedence. I don't think it's great practice tho, I'd love us to not perpetuate it. 🤷🏻‍♂️

File: sdk/common/src/main/java/io/opentelemetry/sdk/common/InstrumentationScopeInfo.java
Comment (jack-berg) [inline sdk/common/src/main/java/io/opentelemetry/sdk/common/InstrumentationScopeInfo.java:34]: I'll deprecate this in a followup PR. Doing it touches 50+ files so better keep that as a dedicated refactor.

File: sdk/common/src/test/java/io/opentelemetry/sdk/internal/ComponentRegistryTest.java
Comment (jack-berg) [reply sdk/common/src/test/java/io/opentelemetry/sdk/internal/ComponentRegistryTest.java]: > I don't think it's great practice tho

I actually prefer grouping similar tests into a single test function. It can make it really easy to succinctly express the intended behavior the code. In certain cases, splitting each scenario into a separate adds keystrokes and makes the code harder to understand. Illegal arguments is one good example, and here's another [example](https://github.com/open-telemetry/opentelemetry-java/pull/4700/files#diff-301b1eb1bf3254c1c1ecaca01c8377184accbed10cc6030fa44541bf27848099R45-R76).

Comment (breedx-splk) [reply sdk/common/src/test/java/io/opentelemetry/sdk/internal/ComponentRegistryTest.java]: Yup, like all things -- pros and cons. You forgot that separate methods runs slower too. 😁 
When something fails in a test that's testing several things, I think it becomes much harder to find out which condition is at fault. Furthermore, the first failure might hide the fact that several conditions are broken.

**Comment by jack-berg** [comment]
This is ready to merge IMO, waiting for an approval from @jkwatson since this impacts our public APIs.

**Review by jkwatson — APPROVED** [review]
thanks!

**Pull request merged**
